### PR TITLE
Allow undo of article fragment remove

### PR DESCRIPTION
**Pull request by c-Rolland**
## Subject

I am targeting this branch, because i propose a new article fragments deletion feature. This feature  move the removed fragment input's in another dom and this allows the cancellations of removed fragments. 

This feature fix too a bug of the removal of predefined fragments in the creation of new article. Bug detected at Ekino.

## Changelog

```markdown
### Added
- Added to fragmentList.jquery.js `moveSelectedFragmentOutsideOfForm()` to move selected fragment in tmp dom
- Added to fragmentList.jquery.js `cancelFragmentDeletion()`  to cancel the fragment deletion

### Changed
The 'removed' fragments are not removed completely just moved in another dom, 
which allows the cancellation. 
         
```

### CODE REVIEW

File: src/Resources/public/js/fragmentList.jquery.js [hunk 1]
```diff
@@ -184,31 +188,91 @@
...
-                this.setActive(this.$list.children(':first-child'));
+               this.setActive(this.$list.children(':first-child'));
             }
 
             // Reorder
             this.resetOrder();
+        },
+
+        /**
+         * This method move the selected fragment and from elements in div#deletedFragments
```
Comment (mazsudo) [inline src/Resources/public/js/fragmentList.jquery.js]: `form elements` i guess ;)

File: src/Resources/public/js/fragmentList.jquery.js [hunk 2]
```diff
@@ -184,31 +188,91 @@
...
+            var fragmentTitle = fragmentElement.find('.fragment__title').text();
+
+            var cancelIcon = $('<i>');
+            cancelIcon.attr('class', 'fa fa-repeat');
+            cancelIcon.on('click', this.cancelFragmentDeletion);
+
+            resetRemovedElement.append(cancelIcon);
+            resetRemovedElement.append(' <i>' +  fragmentTitle + '</i>');
+
+            // Insert the title of fragment and the reset button at the pace of removed fragment.
```
Comment (mazsudo) [inline src/Resources/public/js/fragmentList.jquery.js]: `Insert the fragment's title and the reset button in the place of removed fragment.`?

**Review by mazsudo — CHANGES_REQUESTED** [review]

**Comment by c-Rolland** [comment]
@bmaziere hello ^^ we need this changes at Ekino ^^ can you review the added code pls ? ^^

**Comment by core23** [comment]
Please run cs fixer

**Comment by OskarStark** [comment]
@core23 the PHP CS issues should be fixed by a rebase

File: src/Resources/public/js/fragmentList.jquery.js [hunk 3]
```diff
@@ -37,6 +37,10 @@
             this.build();
             this.setEvents();
             this.setDraggableElements();
+            var divElement = $('<div>');
+            divElement.attr('id', 'deletedFragments');
```
Comment (mazsudo) [inline src/Resources/public/js/fragmentList.jquery.js]: think we should follow some kind of convention here ;) id should be hypens separated as mainly in sonata admin templates

**Comment by OskarStark** [comment]
I would appreciate some docs for this new feature, or at least a note in the documentation

**Review by mazsudo — CHANGES_REQUESTED** [review]

**Comment by c-Rolland** [comment]
> I would appreciate some docs for this new feature, or at least a note in the documentation

@OskarStark i can add it in witch doc file?

**Comment by mazsudo** [comment]
any news on this @c-Rolland ? i guess you can add a new dedicated doc file as it is a new feature

**Comment by c-Rolland** [comment]
@bmaziere the new doc file about the new feat its created .. can you check it pls ? ^^

File: docs/reference/article_fragment.rst [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Article's Fragment
+==================
+
+This feature move the removed fragment input's in another DOM and thereafter
```
Comment (mazsudo) [inline docs/reference/article_fragment.rst]: moves

File: docs/reference/article_fragment.rst [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+this allows the cancellations of removed fragments.
+After the deletion of a fragment, each deleted fragment item in the fragment list
+contain a reset button (to undo the deletion) and the fragment title. This action triggers undo of
```
Comment (mazsudo) [inline docs/reference/article_fragment.rst]: contains

File: docs/reference/article_fragment.rst [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+the deletion of the fragment so the moved fragment input's (and elements) are moving in their original place.
+
+.. note::
+    This feature fix too a bug of the removal of predefined fragments in the creation of new article.
```
Comment (mazsudo) [inline docs/reference/article_fragment.rst]: not sure this note is necessary as soon as the previous bug is resolved ;)

File: docs/reference/article_fragment.rst [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+
```
Comment (mazsudo) [inline docs/reference/article_fragment.rst]: this file should only concern fragment deletion so should be named delete_fragment.rst IMO. It has to be referenced in index.rst file (will fix a part of the build) (and you should rebase ;) will fix the second part of the build 
 i mean flintCI part)

**Review by mazsudo — CHANGES_REQUESTED** [review]

**Review by mazsudo — DISMISSED** [review]

File: docs/reference/delete_fragment.rst [hunk 8]
```diff
@@ -0,0 +1,8 @@
+Article's Fragment
+==================
+
+This feature moves the removed fragment input's in another DOM and thereafter
```
Comment (greg0ire) [inline docs/reference/delete_fragment.rst]: :thinking: should "input's" become "inputs"? same below

File: src/Resources/public/js/fragmentList.jquery.js [hunk 9]
```diff
@@ -184,31 +188,91 @@
...
+            // Insert the fragment's title and the reset button in the place of removed fragment.
+            fragmentElement.empty();
+            fragmentElement.append(resetRemovedElement);
+
+            // Insert the selected fragment and form into #deleted-fragments.
+            $('#deleted-fragments').append(divElement);
+        },
 
+        /**
+         * This method cancel the fragment deletion.
```
Comment (greg0ire) [inline src/Resources/public/js/fragmentList.jquery.js]:
```suggestion
         * This method cancels the fragment deletion.
```

File: src/Resources/public/js/fragmentList.jquery.js [hunk 10]
```diff
@@ -259,7 +323,6 @@
          */
         build: function() {
             var self = this;
-
```
Comment (greg0ire) [inline src/Resources/public/js/fragmentList.jquery.js:262]: Why did you remove this line?

File: src/Resources/public/js/fragmentList.jquery.js [hunk 11]
```diff
@@ -184,31 +188,91 @@
...
-                this.setActive(this.$list.children(':first-child'));
+               this.setActive(this.$list.children(':first-child'));
             }
 
             // Reorder
             this.resetOrder();
+        },
+
+        /**
+         * This method move the selected fragment and form elements in div#deleted-fragments
```
Comment (greg0ire) [inline src/Resources/public/js/fragmentList.jquery.js]:
```suggestion
         * This method moves the selected fragment and form elements in div#deleted-fragments
```

File: src/Resources/public/js/fragmentList.jquery.js [hunk 12]
```diff
@@ -184,31 +188,91 @@
...
+         *
+         * @param id
+         * @param fragmentElement
+         * @param formElement
```
Comment (greg0ire) [inline src/Resources/public/js/fragmentList.jquery.js]: What is the purpose of these docblocks? Shouldn't there be types?

File: docs/reference/delete_fragment.rst [hunk 13]
```diff
@@ -0,0 +1,8 @@
+Article's Fragment
+==================
+
+This feature moves the removed fragment inputs in another DOM and thereafter
+this allows the cancellations of removed fragments.
+After the deletion of a fragment, each deleted fragment item in the fragment list
+contains a reset button (to undo the deletion) and the fragment title. This action triggers undo of
+the deletion of the fragment so the moved fragment input's (and elements) are moving in their original place.
```
Comment (mazsudo) [inline docs/reference/delete_fragment.rst]: the deletion of the fragment so the deleted fragment inputs (and elements) are moved to their original place.

**Review by mazsudo — CHANGES_REQUESTED** [review]

**Review by mazsudo — APPROVED** [review]

**Comment by greg0ire** [comment]
@c-Rolland did you remove your repository or something? It says `unknown_repository` and I cannot use the code suggestion feature.

**Comment by c-Rolland** [comment]
@greg0ire https://github.com/sonata-project/SonataArticleBundle/pull/140 ^^

**Pull request closed without merging**